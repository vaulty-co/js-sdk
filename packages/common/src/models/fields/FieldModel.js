import { immerable } from 'immer';
import { uniqueId } from '@js-sdk/common/src/helpers/uniqueId';
import { pick } from '@js-sdk/common/src/helpers/pick';

import {
  FIELD_STATUSES,
  INITIAL_FIELD_STATUS,
} from './constants';
import {
  updateSettings,
} from './updaters/updateSettings';

const ALLOWED_MODEL_JSON_PROPERTIES = ['id', 'type', 'status', 'settings'];

/**
 * Field's model
 * @class
 */
class FieldModel {
  static get STATUSES() {
    return FIELD_STATUSES;
  }

  /**
   * Create FiledModel by its JSON
   * @param {FieldModelJSON} modelJSON
   */
  static of(modelJSON) {
    /**
     * @type {FieldModelOptions}
     */
    const fieldModelOptions = pick(modelJSON, ALLOWED_MODEL_JSON_PROPERTIES);
    return new FieldModel(fieldModelOptions);
  }

  /**
   * @param {FieldModelOptions} [options = {}]
   */
  constructor(options = {}) {
    this[immerable] = true;

    this.id = options.id || uniqueId();
    this.type = options.type || 'unknown';
    this.status = options.status || INITIAL_FIELD_STATUS;
    this.setSettings(options.settings);
  }

  /**
   * Setup new status for field
   * @param {FieldStatus} newStatus
   */
  setStatus(newStatus) {
    this.status = {
      ...this.status,
      ...newStatus,
    };
  }

  /**
   * Set new settings for field
   * @param {FieldSettings} [newSettings = {}]
   */
  setSettings(newSettings = {}) {
    this.settings = updateSettings(this.settings, newSettings);
    this.setStatus({
      enabling: (
        this.settings.disabled
          ? FIELD_STATUSES.ENABLING.DISABLED
          : FIELD_STATUSES.ENABLING.ENABLED
      ),
    });
  }

  /**
   * Get field name
   * @returns {?string}
   */
  getName() {
    return this.settings?.name;
  }

  /**
   * Get field style
   * @param {string} [styleName]
   * @returns {string|Object}
   */
  getStyle(styleName) {
    const style = this.settings?.style ?? {};
    if (!styleName) {
      return style;
    }
    return style[styleName] || '';
  }

  /**
   * Get field validators
   * @returns {Array<ValidatorName>}
   */
  getValidators() {
    return this.settings?.validators ?? [];
  }

  /**
   * Get field attributes
   * @returns {FieldAttributes}
   */
  getAttributes() {
    return {
      disabled: this.settings?.disabled ?? false,
    };
  }

  /**
   * Prepare a plain object for using in JSON.stringify()
   * @returns {FieldModelJSON}
   */
  toJSON() {
    return pick(this, ALLOWED_MODEL_JSON_PROPERTIES);
  }
}

export default FieldModel;
export {
  FieldModel,
};
