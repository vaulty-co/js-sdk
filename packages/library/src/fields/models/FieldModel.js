import { immerable } from 'immer';
import { uniqueId } from '@js-sdk/utils/src/helpers/uniqueId';

import {
  FIELD_NODE_STATUSES,
  FIELD_CONTENT_STATUSES,
  FIELD_VALIDATION_STATUSES,
  FIELD_READINESS_STATUSES,
  FIELD_FOCUS_STATUSES,
} from '../constants';

/**
 * @typedef {Object} FieldStatus
 * @property {FIELD_NODE_STATUSES} [node]
 * @property {FIELD_CONTENT_STATUSES} [content]
 * @property {FIELD_VALIDATION_STATUSES} [validation]
 * @property {FIELD_READINESS_STATUSES} [readiness]
 * @property {FIELD_FOCUS_STATUSES} [focus]
 */

/**
 * @typedef {Object} FieldModelOptions
 * @property {string} [type='unknown'] - type of field
 */

/**
 * Field's model
 * @class
 */
class FieldModel {
  /**
   * @param {FieldModelOptions} [options = {}]
   */
  constructor(options = {}) {
    this[immerable] = true;

    /**
     * @type {string}
     */
    this.id = uniqueId();
    /**
     * @type {string}
     */
    this.type = options?.type ?? 'unknown';
    /**
     * @type {FieldStatus}
     */
    this.status = {
      node: FIELD_NODE_STATUSES.UNMOUNTED,
      content: FIELD_CONTENT_STATUSES.EMPTY,
      validation: FIELD_VALIDATION_STATUSES.UNKNOWN,
      readiness: FIELD_READINESS_STATUSES.READY,
      focus: FIELD_FOCUS_STATUSES.UNFOCUSED,
    };
  }

  /**
   * Setup new status for field
   * @param {FieldStatus} newStatus
   */
  setStatus(newStatus) {
    this.status = {
      ...this.status,
      ...newStatus,
    };
  }
}

export default FieldModel;
export {
  FieldModel,
};
