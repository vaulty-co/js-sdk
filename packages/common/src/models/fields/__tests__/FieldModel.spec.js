import { FieldModel } from '../FieldModel';
import {
  FIELD_READINESS_STATUSES,
} from '../constants';

describe('FieldModel', () => {
  describe('#constructor', () => {
    it('should work fine, when options is not specified', () => {
      expect(() => {
        new FieldModel();
      }).not.toThrowError();
    });

    it('should setup type as "unknown", if it is not specified', () => {
      const field = new FieldModel({});

      expect(field.type).toBe('unknown');
    });
  });

  describe('#setStatus', () => {
    it('should set status to with replacing only set statuses', () => {
      const field = new FieldModel();

      field.setStatus({
        readiness: FIELD_READINESS_STATUSES.LOADING,
      });

      expect(field.status).toMatchSnapshot();
    });
  });
});
