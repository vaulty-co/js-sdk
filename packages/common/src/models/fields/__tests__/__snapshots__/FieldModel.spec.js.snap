// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldModel #setStatus should set status to with replacing only set statuses 1`] = `
Object {
  "content": "empty",
  "enabling": "enabled",
  "focus": "unfocused",
  "node": "unmounted",
  "readiness": "loading",
  "validation": Object {
    "invalidValidators": Array [],
    "status": "unknown",
  },
}
`;
